### PR TITLE
Add managed disk option for data disks

### LINKED ISSUE #964: Add managed disk options to data disks
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Support optionally specifying ManagedDisk storage type for data disks https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/types.go#L330:6, reusing the same type as OSDisks https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/types.go#L323.

This is a good end-to-end feature for anyone looking to get familiar with CAPZ, and I'm happy to provide further guidance for anyone who wants to pick it up.

The solution should include:

- API changes 
- Update to the VM and VMSS services to set the options when creating the VM/VMSS
- Webhook validation
- Unit tests
- Template changes and/or documentation updates (data-disks.md)

Don't know where to start? Search "StorageAccountType" in the codebase to see how it's currently implemented for OS Disks.

/help
/good-first-issue

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by gdhuper**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add support for ManagedDisk storage type for data disks.

**Which issue(s) this PR fixes**:
Fixes #964 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] API changes
- [x] Update to the VM and VMSS services to set the options when creating the VM/VMSS
- [x] Webhook Validation
- [x] Template changes and/or documentation updates (data-disks.md)
- [x] includes documentation
- [x] adds unit tests / e2e tests 
- [x] squashed commits

**Release note**:

```release-note
Added option for managed disks for data disks.
```

### CODE REVIEW

**Comment by nader-ziada** [comment]
this PR is re-doing #1150

File: api/v1alpha3/azuremachine_validation.go [hunk 1]
```diff
@@ -165,6 +174,26 @@ func ValidateManagedDisk(old, new ManagedDisk, fieldPath *field.Path) field.Erro
...
+	allErrs := field.ErrorList{}
+
+	if len(old) != len(new) {
+		allErrs = append(allErrs, field.Invalid(fieldPath, new, "changing data disks after machine creation is not allowed"))
+		return allErrs
+	}
+
+	for i, disk := range new {
+		if disk.ManagedDisk != nil && old[i].ManagedDisk != nil {
+			if errs := ValidateManagedDisk(*old[i].ManagedDisk, *disk.ManagedDisk, fieldPath.Child("managedDisk")); len(errs) > 0 {
```
Comment (shysank) [inline api/v1alpha3/azuremachine_validation.go:188]: Do we expect old and new disk slices to be in the same order? Maybe map it to dictionary with `NameSuffix` as key, and then check for equality?

File: api/v1alpha3/azuremachine_webhook.go [hunk 2]
```diff
@@ -105,6 +105,10 @@ func (m *AzureMachine) ValidateUpdate(oldRaw runtime.Object) error {
 		allErrs = append(allErrs, errs...)
 	}
 
+	if errs := ValidateManagedDisks(old.Spec.DataDisks, m.Spec.DataDisks, field.NewPath("dataDisks")); len(errs) > 0 {
```
Comment (shysank) [inline api/v1alpha3/azuremachine_webhook.go:108]: Can we add a test for this?

Comment (shysank) [reply api/v1alpha3/azuremachine_webhook.go:108]: Also, there is already a `ValidateDataDisks` which validates `spec.DataDisks`. Maybe update that to include `ValidateManagedDisks` check?

File: azure/services/virtualmachines/virtualmachines.go [hunk 3]
```diff
@@ -398,6 +398,16 @@ func (s *Service) generateStorageProfile(ctx context.Context, vmSpec azure.VMSpe
 			Name:         to.StringPtr(azure.GenerateDataDiskName(vmSpec.Name, disk.NameSuffix)),
 			Caching:      compute.CachingTypes(disk.CachingType),
 		}
+
+		if disk.ManagedDisk != nil {
```
Comment (shysank) [inline azure/services/virtualmachines/virtualmachines.go:402]: Same comment as above for tests.

File: azure/services/scalesets/scalesets.go [hunk 4]
```diff
@@ -537,6 +537,16 @@ func (s *Service) generateStorageProfile(vmssSpec azure.ScaleSetSpec, sku resour
 			Lun:          disk.Lun,
 			Name:         to.StringPtr(azure.GenerateDataDiskName(vmssSpec.Name, disk.NameSuffix)),
 		}
+
+		if disk.ManagedDisk != nil {
```
Comment (shysank) [inline azure/services/scalesets/scalesets.go:541]: can we add a test for this or at least modify an existing test to check if the `storageProfile` values are set properly?

**Review by shysank — CHANGES_REQUESTED** [review]

File: api/v1alpha3/azuremachine_validation.go (hunk 1 above)
Comment (gdhuper) [reply api/v1alpha3/azuremachine_validation.go:188]: Good call out. I will double check and use a dictionary

File: api/v1alpha3/azuremachine_validation.go [hunk 5]
```diff
@@ -165,6 +174,26 @@ func ValidateManagedDisk(old, new ManagedDisk, fieldPath *field.Path) field.Erro
 	return allErrs
 }
 
+// ValidateManagedDisks validates updates to Data disks' managed disk field.
+func ValidateManagedDisks(old, new []DataDisk, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(old) != len(new) {
```
Comment (gdhuper) [inline api/v1alpha3/azuremachine_validation.go:181]: @CecileRobertMichon I am not sure if this check is required anymore. Can you clarify what changes are allowed in terms of updates to data disks?  Can a user add/remove data disks? what specific fields for data disks can a user update vs which fields are read only?

File: api/v1alpha3/azuremachine_webhook.go (hunk 2 above)
Comment (gdhuper) [reply api/v1alpha3/azuremachine_webhook.go]: After trying out a few possibilities, I think it's best if we have this as a separate function as we are only validating managed disks for data disks. If we want to incorporate this in `ValidateDataDisks`, the function signature would change and would not be consistent with `ValidateOsDisk`. Any ideas/suggestions here?

Comment (shysank) [reply api/v1alpha3/azuremachine_webhook.go]: I thought it was slightly confusing to have two validation methods for the same field. I don't mind changing the signature if needed. If you think changes are going to be big, I'm okay with just adding comments for now, and doing it in a separate refactoring pr.

File: api/v1alpha3/azuremachine_validation.go (hunk 5 above)
Comment (CecileRobertMichon) [reply api/v1alpha3/azuremachine_validation.go:181]: Currently, a user connot add/remove/modify data disks on an existing machine. Validation should reflect that.

**Comment by CecileRobertMichon** [comment]
@gdhuper please squash commits 

lgtm
/assign @nader-ziada

File: api/v1alpha3/azuremachine_validation.go (hunk 5 above)
Comment (gdhuper) [reply api/v1alpha3/azuremachine_validation.go:181]: Thank you for the clarification.

**Comment by shysank** [comment]
/lgtm

@gdhuper Thanks for adding tests!

**Comment by gdhuper** [comment]
> /hold for rebase

Do I need to rebase with master? Or just squash commits?

**Comment by CecileRobertMichon** [comment]
@gdhuper I meant rebase to squash the commits

You can do that with:

```
git fetch upstream
git rebase -i upstream/master
<edit the commits in vim to squash commit 2 and 3>
git push -f
```
assuming you have a remote for upstream called "upstream"

**Comment by gdhuper** [comment]
> @gdhuper I meant rebase to squash the commits
> 
> 
> 
> You can do that with:
> 
> 
> 
> ```
> 
> git fetch upstream
> 
> git rebase -i upstream/master
> 
> <edit the commits in vim to squash commit 2 and 3>
> 
> git push -f
> 
> ```
> 
> assuming you have a remote for upstream called "upstream"

Got it. Thanks

File: api/v1alpha3/azuremachine_webhook.go (hunk 2 above)
Comment (gdhuper) [reply api/v1alpha3/azuremachine_webhook.go]: Makes sense. I changed the function name to `ValidateDataDisksUpdate` to make it less confusing and added more checks. Thank you for your feedback.

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve

**Comment by gdhuper** [comment]
> /retest

@CecileRobertMichon Is there something I can do on my part to fix this test?

**Comment by CecileRobertMichon** [comment]
> You mentioned that this is a known flake, but is there something I can do on my part to fix this test?

It's unrelated to your PR but if you'd like to help debug it, you can start by looking at the [artifacts](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1163/pull-cluster-api-provider-azure-e2e-windows/1361734325753614336/) and see if you find anything interesting in logs. If it's not already a cause being tracked in an issue, we need to open a new issue describing the problem with links to the logs. The most common one that I've seen recently that I've been meaning to investigate is #1114

**Pull request merged**